src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5275,11 +5275,39 @@ private ScanTypeArgumentListKind ScanTypeArgumentList(NameOptions options)
                 return ScanTypeArgumentListKind.DefiniteTypeArgumentList;
             }
 
-            if (!this.ScanPossibleTypeArgumentList())
+            // We're in an expression context, and we have a < token.  This could be a 
+            // type argument list, or it could just be a relational expression.  
+            //
+            // Scan just the type argument list portion (i.e. the part from < to > ) to
+            // see what we think it could be.  This will give us one of three possibilities:
+            //
+            //      result == ScanTypeFlags.NotType.
+            //
+            // This is absolutely not a type-argument-list.  Just return that result immediately.
+            //
+            //      result != ScanTypeFlags.NotType && isDefinitelyTypeArgumentList.
+            //
+            // This is absolutely a type-argument-list.  Just return that result immediately
+            // 
+            //      result != ScanTypeFlags.NotType && !isDefinitelyTypeArgumentList.
+            //
+            // This could be a type-argument list, or it could be an expression.  Need to see
+            // what came after the last '>' to find out which it is.
+
+            SyntaxToken lastTokenOfList = null;
+            ScanTypeFlags possibleTypeArgumentFlags = ScanPossibleTypeArgumentList(
+                ref lastTokenOfList, out bool isDefinitelyTypeArgumentList);
```

Comment:
note: i'm not a huge fan of having multiple results.  But i couldn't find a great way to thread the bool through the result value.  If anyone has any suggestions on the best way to do this, i'm all ears.  Specifically, i didn't want people to have to try to figure out things like "PossibleGenericTypeOrMethod" and "MustBeGenericTypeOrMethod".  But maybe that's appropriate?